### DISCUSSION TITLE
Does printer redirection from Linux to Linux?

### DISCUSSION THREAD

**Discussion by logmestereN** [FreeRDP/FreeRDP#10300] [Q&A] [answered]
Hello! I'm trying to redirect printers (no matter local or network printers) via freerdp from ubuntu 22.04 to ubuntu 22.04. Redirection works fine when i try it from ubuntu to windows, but doesnt work on ubuntu-ubuntu. 

Tried to specify exact printer with driver and tried to just `/printer`, but `lpstat -v` shows "no destinations added".

I'm using FreeRDP version 3.5.2-dev0, also tried on 2.6.1. Buildinfo:
`Build configuration: BUILD_TESTING=OFF WINPR_HAVE_AIO_H=1 WINPR_HAVE_EXECINFO_BACKTRACE=1 WINPR_HAVE_EXECINFO_BACKTRACE_SYMBOLS=1 WINPR_HAVE_EXECINFO_BACKTRACE_SYMBOLS_FD=1 WINPR_HAVE_EXECINFO_HEADER=1 WINPR_HAVE_FCNTL_H=1 WINPR_HAVE_GETLOGIN_R=1 WINPR_HAVE_GETPWUID_R=1 WINPR_HAVE_INTTYPES_H=1 WINPR_HAVE_POLL_H=1 WINPR_HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 WINPR_HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS= WINPR_HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL=1 WINPR_HAVE_STDBOOL_H=1 WINPR_HAVE_STDINT_H=1 WINPR_HAVE_STRNDUP=1 WINPR_HAVE_SYSLOG_H=1 WINPR_HAVE_SYS_EVENTFD_H=1 WINPR_HAVE_SYS_FILIO_H= WINPR_HAVE_SYS_SELECT_H=1 WINPR_HAVE_SYS_SOCKIO_H= WINPR_HAVE_SYS_TIMERFD_H=1 WINPR_HAVE_TM_GMTOFF=1 WINPR_HAVE_UNISTD_H=1 WINPR_HAVE_UNWIND_H=1 WITH_AAD=ON WITH_ABSOLUTE_PLUGIN_LOAD_PATHS=ON WITH_ADD_PLUGIN_TO_RPATH=OFF WITH_ALSA=ON WITH_BINARY_VERSIONING=OFF WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CJSON_REQUIRED=OFF WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CLIENT_SDL=ON WITH_CLIENT_SDL2=ON WITH_CLIENT_SDL3=OFF WITH_CLIENT_SDL_AVAILABLE=1 WITH_CLIENT_SDL_VERSIONED=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_CODECS=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_EVENTS=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SCHANNEL=OFF WITH_DEBUG_SDL_EVENTS=OFF WITH_DEBUG_SDL_KBD_EVENTS=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=ON WITH_DSP_FFMPEG_AVAILABLE=1 WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=ON WITH_FREERDP_DEPRECATED=OFF WITH_FREERDP_DEPRECATED_COMMANDLINE=OFF WITH_FULL_CONFIG_PATH=OFF WITH_FUSE=ON WITH_GFX_FRAME_DUMP=OFF WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=OFF WITH_INTERNAL_MD4=OFF WITH_INTERNAL_MD5=OFF WITH_INTERNAL_RC4=OFF WITH_JPEG=OFF WITH_JSONC_REQUIRED=OFF WITH_JSON_DISABLED=OFF WITH_KRB5=ON WITH_KRB5_NO_NTLM_FALLBACK=OFF WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBRESSL=OFF WITH_LODEPNG=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_NATIVE_SSPI=OFF WITH_NEON=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSSL=ON WITH_OPUS=OFF WITH_OSS=ON WITH_PCSC=ON WITH_PKCS11=ON WITH_PLATFOM_SERVER=OFF WITH_PLATFORM_SERVER=ON WITH_POLL=ON WITH_PROFILER=OFF WITH_PROXY=ON WITH_PROXY_APP=ON WITH_PROXY_EMULATE_SMARTCARD=OFF WITH_PROXY_MODULES=ON WITH_PULSE=OFF WITH_RDTK=ON WITH_SAMPLE=ON WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SDL_IMAGE_DIALOGS=OFF WITH_SDL_LINK_SHARED=ON WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SMARTCARD_EMULATE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SMARTCARD_PCSC=ON WITH_SOXR=OFF WITH_SSE2=OFF WITH_SWSCALE=OFF WITH_SYSTEMD=ON WITH_THIRD_PARTY=OFF WITH_TIMEZONE_ICU=OFF WITH_UNICODE_BUILTIN=OFF WITH_URIPARSER=OFF WITH_VAAPI=OFF WITH_VAAPI_AVAILABLE=1 WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_VERBOSE_WINPR_ASSERT=ON WITH_VIDEO_FFMPEG=ON WITH_VIDEO_FFMPEG_AVAILABLE=1 WITH_WAYLAND=ON WITH_WEBVIEW=OFF WITH_WINPR_DEPRECATED=OFF WITH_WINPR_JSON=ON WITH_WINPR_TOOLS=ON WITH_WIN_CONSOLE=ON WITH_X11=ON WITH_XCURSOR=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XV=ON
Build type:          Release
CFLAGS:               -Wall -Wpedantic -Wno-padded -Wno-cast-align -Wno-declaration-after-statement -fPIC -Wall -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -fno-omit-frame-pointer
Compiler:            GNU, 11.4.0
Target architecture: x64`

Same builds on client and server.

CUPS is enabled, cupsd.conf below:

> LogLevel warn
> 
> PageLogFormat
> MaxLogSize 0
> ErrorPolicy retry-job
> Port 631
> Listen /run/cups/cups.sock
> Browsing On
> BrowseLocalProtocols all
> DefaultAuthType Basic
> WebInterface Yes
> IdleExitTimeout 60
> <Location />
>   Order allow,deny
>   Allow all
> </Location>
> <Location /admin>
>   Order allow,deny
>   Allow all
> </Location>
> <Location /admin/conf>
>   AuthType Default
>   Require user @SYSTEM
>   Order allow,deny
>   Allow all
> </Location>
> <Location /admin/log>
>   AuthType Default
>   Require user @SYSTEM
>   Order allow,deny
> </Location>
> <Policy default>
>   JobPrivateAccess default
>   JobPrivateValues default
>   SubscriptionPrivateAccess default
>   SubscriptionPrivateValues default
>   <Limit Create-Job Print-Job Print-URI Validate-Job>
>     Order deny,allow
>   </Limit>
>   <Limit Send-Document Send-URI Hold-Job Release-Job Restart-Job Purge-Jobs Set-Job-Attributes Create-Job-Subscription Renew-Subscri>
>     Require user @OWNER @SYSTEM
>     Order deny,allow
>   </Limit>
>   <Limit CUPS-Get-Document>
>     AuthType Default
> 

So there is a question: does printer redirection works from linux to linux?

Also additional question: is it possible to redirect port via rdp2tcp from linux to linux or it only works for linux-windows? If there is  another built-in/not built-in solution to redirect printers or ports let me know please. Thanks in advance!

**Comment by akallabeth** [comment, marked as answer]
@logmestereN that depends on what you use server side. You need to check if your software supports that. 
some options available are:

* [xrdp](https://github.com/neutrinolabs/xrdp)
* [gnome-remote-desktop](https://gitlab.gnome.org/GNOME/gnome-remote-desktop) (`FreeRDP` based)
* [ogon](https://github.com/ogon-project/ogon) (`FreeRDP` based)

there are some others too, but these should cover most

**Reply by logmestereN** [reply]
First two cant redirect printers and ogon isnt up-to-date, have problems with building it properly, but i'll try further

Will be thankful if you answer: can freerdp work as a server? Tried with freerdp-shadow-cli but this seems cant provide full remote desktop environment

And is there another solutions to forward ports in freerdp? Except rdp2tcp cause its seems like server side compiles for windows only

**Reply by akallabeth** [reply]
ogon requires `FreeRDP 2`, but builds with it. (requires a `X11` patch though, that is annoying yes.

In the `FreeRDP` repo we only have the shadow server (mirror a user session) and that is quite limited (more or less just a proof of concept)

as for `rdp2tcp`  I don´t think g-r-d and xrdp allow loading libraries as channels in a user session (might be wrong), but you most likely need to patch them

**Reply by logmestereN** [reply]
Thank you for answer! Will continue my research with ogon and wait for updates of freerdp

**Reply by logmestereN** [reply]
> (requires a X11 patch though, that is annoying yes.


Thats is really annoying
@akallabeth Did you try to patch it? Maybe you have a link for guide how to patch it? Sorry if my questions a bit off-topic
